### PR TITLE
Add cache by session function

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
At current time we have a lot of values that cached in local variables. This function will allow to make this code more friendly.

I expect first usages will be introduced in this PR https://github.com/freeorion/freeorion/pull/576

### CODE REVIEW

File: default/AI/freeorion_tools.py [hunk 1]
```diff
@@ -132,6 +132,25 @@ def chat_human(message):
     print "\nChat Message to human: %s\n" % remove_tags(message)
 
 
+def cache_by_session(function):
+    """
+    Cache a function value by session.
+    Wrap only functions with hashable arguments.
```
Comment (Morlic-fo) [inline default/AI/freeorion_tools.py]: Wraps

**Comment by Cjkjvfnby** [comment]
Fixed.

Github have new feature squash and rebase instead of merge, it perfectly suits for this PR

**Comment by Morlic-fo** [comment]
> Github have new feature squash and rebase instead of merge, it perfectly suits for this PR

I still only have the option to do a classic merge.

**Comment by MatGB** [comment]
Same here, it needs to be enabled by an admin, I've commented on the thread discussing it that it should be an option that we should use rarely, there are times when I'd much prefer to see a single commit at the time it went in.

**Comment by Cjkjvfnby** [comment]
Looks like it is not disabled for me. 
- click Merge pull request.
- click arrow near `Confirm merge`
- select squash and merge
- edit commit message

**Comment by Vezzra** [comment]
The option has been enabled by default when github introduced the feature, no one has turned it off as AFAICT. @MatGB, @Morlic-fo, apparently you didn't realize that you need to click "Merge pull request" first, and only then get the option to choose...?

And apparently @Morlic-fo figured it out after @Cjkjvfnby comment and did a "squash and merge" with this commit, right?

**Comment by MatGB** [comment]
Pretty much, yes, thanks for checking. Every time I think I've learnt what I need to know about git/github something changes ;-)

**Comment by Morlic-fo** [comment]
> @MatGB, @Morlic-fo, apparently you didn't realize that you need to click "Merge pull request" first, and only then get the option to choose...?

Yep, that was it. I expected to have the dropdown already available from the start...

**Pull request merged**